pkg/boot/survey/survey.go

```diff
@@ -0,0 +1,367 @@
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	capnp "capnproto.org/go/capnp/v3"
+	"github.com/cstockton/go-conv"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	"github.com/libp2p/go-libp2p-core/record"
+	cpSurvey "github.com/wetware/casm/internal/api/survey"
+)
+
+const (
+	discLimit = 10
+	discTTL   = time.Minute
+	discDist  = uint8(255)
+
+	timeout         = 5 * time.Second
+	maxDatagramSize = 8192
+)
+
+type Survey struct {
+	h             host.Host
+	mustFind      map[string]chan peer.AddrInfo
+	mustAdvertise map[string]chan time.Duration
+	mu            sync.Mutex
+
+	c   comm
+	err atomic.Value
+}
+
+func NewSurvey(h host.Host, addr net.Addr, opt ...Option) (surv *Survey, err error) {
+	config := Config{}
+	config.Apply(opt)
+
+	lconn, err := config.Listen(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	dconn, err := config.Dial(addr)
+	if err != nil {
+		lconn.Close()
+		return nil, err
+	}
+
+	c := comm{addr: addr, lconn: lconn, dconn: dconn, cherr: make(chan error)}
+
+	surv = &Survey{h: h, mustFind: make(map[string]chan peer.AddrInfo),
+		mustAdvertise: make(map[string]chan time.Duration), c: c}
+
+	go c.Listen(surv.multicastHandler)
+
+	go func() {
+		for {
+			err := <-c.cherr
+			if err != nil {
+				surv.err.Store(err)
+			} else {
+				surv.err.Store(fmt.Errorf("closed"))
+			}
+			lconn.Close()
+			dconn.Close()
+			return
+		}
+	}()
+
+	return surv, nil
+}
+
+func (surv *Survey) Close() {
+	if err := surv.err.Load(); err == nil {
+		surv.c.Close()
+		for surv.err.Load() == nil {
+		}
+	}
+}
+
+func (surv *Survey) multicastHandler(n int, src net.Addr, buffer []byte) {
+	msg, err := capnp.UnmarshalPacked(buffer[:n])
+	if err != nil {
+		return
+	}
+
+	root, err := cpSurvey.ReadRootSurveyPacket(msg)
+	if err != nil {
+		return
+	}
+
+	switch root.Which() {
+	case cpSurvey.SurveyPacket_Which_request:
+		request, err := root.Request()
+		if err != nil {
+			return
+		}
+		surv.handleRequest(request)
+	case cpSurvey.SurveyPacket_Which_response:
+		response, err := root.Response()
+		if err != nil {
+			return
+		}
+		surv.handleResponse(response)
+	default:
+	}
+}
+
+func (surv *Survey) handleRequest(request cpSurvey.SurveyRequest) {
+	surv.mu.Lock()
+	defer surv.mu.Unlock()
+
+	// validate requester
+	envelope, err := request.Src()
+	if err != nil {
+		return
+	}
+
+	var rec peer.PeerRecord
+	if _, err = record.ConsumeTypedEnvelope(envelope, &rec); err != nil {
+		return
+	}
+	if rec.PeerID == surv.h.ID() {
+		return // request comes from itself
+	}
+
+	if dist([]byte(surv.h.ID()), []byte(rec.PeerID))>>uint32(request.Distance()) != 0 {
+		return
+	}
+
+	ns, err := request.Namespace()
+	if err != nil {
+		return
+	}
+
+	if _, ok := surv.mustAdvertise[ns]; !ok {
+		return
+	}
+
+	response, err := surv.buildResponse(ns)
+	if err == nil {
+		go surv.c.Send(response)
```

Comment:
I think this is a concurrency bug.  Methods on `net.PacketConn` implementations are generally not safe for concurrent access.

The more I stare at this, the more I'm convinced that we should be using channels to orchestrate the flow of messages.  We effectively have a pipeline where the read-conn generates messages and passes them to a handler which (in the case of `Request` packets) passes a `Response` packet to the write-conn.

As such, I think it would make sense to run both `net.PacketConn`s in background goroutines, and have one push incoming packets to a receive-chan and the other consume packets from a send-chan.